Got panic in case of parsing null as did document

**What I'm trying to do**

I'm parsing "null" string by did.ParseDocument

**Expected result**

I'm getting nil, error or empty struct.

**Actual result**

```
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0x13e8130]

goroutine 6 [running]:
testing.tRunner.func1(0xc000154400)
	/usr/local/Cellar/go/1.13.4/libexec/src/testing/testing.go:874 +0x3a3
panic(0x1464900, 0x18609c0)
	/usr/local/Cellar/go/1.13.4/libexec/src/runtime/panic.go:679 +0x1b2
github.com/hyperledger/aries-framework-go/pkg/doc/did.ParseDocument(0xc0001ba818, 0x4, 0x4, 0xc000042760, 0x1083c26, 0x5e625ca9)
	/Users/belezeko/Projects/aries-framework-go/pkg/doc/did/doc.go:401 +0x1a0
github.com/hyperledger/aries-framework-go/pkg/doc/did.TestParseOfNull(0xc000154400)
	/Users/belezeko/Projects/aries-framework-go/pkg/doc/did/doc_test.go:144 +0x57
testing.tRunner(0xc000154400, 0x1503b48)
	/usr/local/Cellar/go/1.13.4/libexec/src/testing/testing.go:909 +0xc9
created by testing.(*T).Run
	/usr/local/Cellar/go/1.13.4/libexec/src/testing/testing.go:960 +0x350

Process finished with exit code 1
```

**Sample code / test case**

```
func TestParseOfNull(t *testing.T) {
	doc, err := ParseDocument([]byte("null"))
	require.NoError(t, err)
	require.Nil(t, doc)
}
```

--- a/pkg/doc/did/doc.go
+++ b/pkg/doc/did/doc.go
@@ -386,16 +386,18 @@ type Proof struct {
 
 // ParseDocument creates an instance of DIDDocument by reading a JSON document from bytes
 func ParseDocument(data []byte) (*Doc, error) {
-	// validate did document
-	if err := validate(data); err != nil {
-		return nil, err
-	}
-
 	raw := &rawDoc{}
 
 	err := json.Unmarshal(data, &raw)
 	if err != nil {
 		return nil, fmt.Errorf("JSON marshalling of did doc bytes bytes failed: %w", err)
+	} else if raw == nil {
+		return nil, errors.New("document payload is not provided")
+	}
+	// validate did document
+	err = validate(data, raw.schemaLoader())
+	if err != nil {
+		return nil, err
 	}
 
 	publicKeys, err := populatePublicKeys(raw.Context[0], raw.PublicKey)
@@ -595,20 +597,18 @@ func decodePK(rawPK map[string]interface{}) ([]byte, error) {
 	return nil, errors.New("public key encoding not supported")
 }
 
-func validate(data []byte) error {
-	// Validate that the DID Document conforms to the serialization of the DID Document data model.
-	// Reference: https://w3c-ccg.github.io/did-spec/#did-documents)
+func (r *rawDoc) schemaLoader() gojsonschema.JSONLoader {
 	schemaLoader := schemaLoaderV1
-
-	var r rawDoc
-	if err := json.Unmarshal(data, &r); err != nil {
-		return fmt.Errorf("unmarshal failed: %w", err)
-	}
-
 	if len(r.Context) > 0 && r.Context[0] == contextV011 {
 		schemaLoader = schemaLoaderV011
 	}
 
+	return schemaLoader
+}
+
+func validate(data []byte, schemaLoader gojsonschema.JSONLoader) error {
+	// Validate that the DID Document conforms to the serialization of the DID Document data model.
+	// Reference: https://w3c-ccg.github.io/did-spec/#did-documents)
 	documentLoader := gojsonschema.NewStringLoader(string(data))
 
 	result, err := gojsonschema.Validate(schemaLoader, documentLoader)

--- a/pkg/doc/did/doc_test.go
+++ b/pkg/doc/did/doc_test.go
@@ -140,6 +140,13 @@ const creator = did + "#key-1"
 const keyType = "Ed25519VerificationKey2018"
 const signatureType = "Ed25519Signature2018"
 
+func TestParseOfNull(t *testing.T) {
+	doc, err := ParseDocument([]byte("null"))
+	require.Error(t, err)
+	require.Nil(t, doc)
+	require.Contains(t, err.Error(), "document payload is not provided")
+}
+
 func TestValid(t *testing.T) {
 	docs := []string{validDoc, validDocV011}
 	for _, d := range docs {
@@ -346,7 +353,7 @@ func TestParseDocument(t *testing.T) {
 	// test error from Unmarshal
 	_, err := ParseDocument([]byte("wrongData"))
 	require.Error(t, err)
-	require.Contains(t, err.Error(), "unmarshal failed")
+	require.Contains(t, err.Error(), "JSON marshalling of did doc bytes bytes failed")
 }
 
 func TestValidateDidDocContext(t *testing.T) {
@@ -358,7 +365,7 @@ func TestValidateDidDocContext(t *testing.T) {
 			raw.Context = nil
 			bytes, err := json.Marshal(raw)
 			require.NoError(t, err)
-			err = validate(bytes)
+			err = validate(bytes, raw.schemaLoader())
 			require.Error(t, err)
 			require.Contains(t, err.Error(), "@context is required")
 		}
@@ -370,7 +377,7 @@ func TestValidateDidDocContext(t *testing.T) {
 		raw.Context = []string{"https://w3id.org/did/v2", "https://w3id.org/did/v1"}
 		bytes, err := json.Marshal(raw)
 		require.NoError(t, err)
-		err = validate(bytes)
+		err = validate(bytes, raw.schemaLoader())
 		require.Error(t, err)
 		require.Contains(t, err.Error(), "Does not match pattern '^https://w3id.org/did/v1$'")
 	})
@@ -385,7 +392,7 @@ func TestValidateDidDocID(t *testing.T) {
 			raw.ID = ""
 			bytes, err := json.Marshal(raw)
 			require.NoError(t, err)
-			err = validate(bytes)
+			err = validate(bytes, raw.schemaLoader())
 			require.Error(t, err)
 			require.Contains(t, err.Error(), "id is required")
 		}
@@ -401,7 +408,7 @@ func TestValidateDidDocPublicKey(t *testing.T) {
 			raw.PublicKey = nil
 			bytes, err := json.Marshal(raw)
 			require.NoError(t, err)
-			err = validate(bytes)
+			err = validate(bytes, raw.schemaLoader())
 			require.NoError(t, err)
 		}
 	})
@@ -414,7 +421,7 @@ func TestValidateDidDocPublicKey(t *testing.T) {
 			delete(raw.PublicKey[0], jsonldID)
 			bytes, err := json.Marshal(raw)
 			require.NoError(t, err)
-			err = validate(bytes)
+			err = validate(bytes, raw.schemaLoader())
 			require.Error(t, err)
 			require.Contains(t, err.Error(), "id is required")
 		}
@@ -428,7 +435,7 @@ func TestValidateDidDocPublicKey(t *testing.T) {
 			delete(raw.PublicKey[0], jsonldType)
 			bytes, err := json.Marshal(raw)
 			require.NoError(t, err)
-			err = validate(bytes)
+			err = validate(bytes, raw.schemaLoader())
 			require.Error(t, err)
 			require.Contains(t, err.Error(), "type is required")
 		}
@@ -440,7 +447,7 @@ func TestValidateDidDocPublicKey(t *testing.T) {
 		delete(raw.PublicKey[0], jsonldController)
 		bytes, err := json.Marshal(raw)
 		require.NoError(t, err)
-		err = validate(bytes)
+		err = validate(bytes, raw.schemaLoader())
 		require.Error(t, err)
 		require.Contains(t, err.Error(), "controller is required")
 	})
@@ -451,7 +458,7 @@ func TestValidateDidDocPublicKey(t *testing.T) {
 		delete(raw.PublicKey[0], jsonldOwner)
 		bytes, err := json.Marshal(raw)
 		require.NoError(t, err)
-		err = validate(bytes)
+		err = validate(bytes, raw.schemaLoader())
 		require.NoError(t, err)
 	})
 
@@ -463,7 +470,7 @@ func TestValidateDidDocPublicKey(t *testing.T) {
 			raw.PublicKey[0]["key1"] = ""
 			bytes, err := json.Marshal(raw)
 			require.NoError(t, err)
-			err = validate(bytes)
+			err = validate(bytes, raw.schemaLoader())
 			require.Error(t, err)
 			require.Contains(t, err.Error(), "Must have at most 4 properties")
 		}
@@ -479,7 +486,7 @@ func TestValidateDidDocAuthentication(t *testing.T) {
 			raw.Authentication = nil
 			bytes, err := json.Marshal(raw)
 			require.NoError(t, err)
-			err = validate(bytes)
+			err = validate(bytes, raw.schemaLoader())
 			require.NoError(t, err)
 		}
 	})
@@ -492,7 +499,7 @@ func TestValidateDidDocAuthentication(t *testing.T) {
 			raw.Authentication[0] = 1
 			bytes, err := json.Marshal(raw)
 			require.NoError(t, err)
-			err = validate(bytes)
+			err = validate(bytes, raw.schemaLoader())
 			require.Error(t, err)
 			require.Contains(t, err.Error(), "Invalid type. Expected: object, given: integer")
 		}
@@ -508,7 +515,7 @@ func TestValidateDidDocAuthentication(t *testing.T) {
 			delete(pk, jsonldID)
 			bytes, err := json.Marshal(raw)
 			require.NoError(t, err)
-			err = validate(bytes)
+			err = validate(bytes, raw.schemaLoader())
 			require.Error(t, err)
 			require.Contains(t, err.Error(), "id is required")
 		}
@@ -524,7 +531,7 @@ func TestValidateDidDocAuthentication(t *testing.T) {
 			delete(pk, jsonldType)
 			bytes, err := json.Marshal(raw)
 			require.NoError(t, err)
-			err = validate(bytes)
+			err = validate(bytes, raw.schemaLoader())
 			require.Error(t, err)
 			require.Contains(t, err.Error(), "type is required")
 		}
@@ -538,7 +545,7 @@ func TestValidateDidDocAuthentication(t *testing.T) {
 		delete(pk, jsonldController)
 		bytes, err := json.Marshal(raw)
 		require.NoError(t, err)
-		err = validate(bytes)
+		err = validate(bytes, raw.schemaLoader())
 		require.Error(t, err)
 		require.Contains(t, err.Error(), "controller is required")
 	})
@@ -551,7 +558,7 @@ func TestValidateDidDocAuthentication(t *testing.T) {
 		delete(pk, jsonldOwner)
 		bytes, err := json.Marshal(raw)
 		require.NoError(t, err)
-		err = validate(bytes)
+		err = validate(bytes, raw.schemaLoader())
 		require.NoError(t, err)
 	})
 
@@ -565,7 +572,7 @@ func TestValidateDidDocAuthentication(t *testing.T) {
 			pk["key1"] = ""
 			bytes, err := json.Marshal(raw)
 			require.NoError(t, err)
-			err = validate(bytes)
+			err = validate(bytes, raw.schemaLoader())
 			require.Error(t, err)
 			require.Contains(t, err.Error(), "Must have at most 4 properties")
 		}
@@ -581,7 +588,7 @@ func TestValidateDidDocService(t *testing.T) {
 			raw.Service = nil
 			bytes, err := json.Marshal(raw)
 			require.NoError(t, err)
-			err = validate(bytes)
+			err = validate(bytes, raw.schemaLoader())
 			require.NoError(t, err)
 		}
 	})
@@ -594,7 +601,7 @@ func TestValidateDidDocService(t *testing.T) {
 			delete(raw.Service[0], jsonldID)
 			bytes, err := json.Marshal(raw)
 			require.NoError(t, err)
-			err = validate(bytes)
+			err = validate(bytes, raw.schemaLoader())
 			require.Error(t, err)
 			require.Contains(t, err.Error(), "id is required")
 		}
@@ -608,7 +615,7 @@ func TestValidateDidDocService(t *testing.T) {
 			delete(raw.Service[0], jsonldType)
 			bytes, err := json.Marshal(raw)
 			require.NoError(t, err)
-			err = validate(bytes)
+			err = validate(bytes, raw.schemaLoader())
 			require.Error(t, err)
 			require.Contains(t, err.Error(), "type is required")
 		}
@@ -622,7 +629,7 @@ func TestValidateDidDocService(t *testing.T) {
 			delete(raw.Service[0], jsonldServicePoint)
 			bytes, err := json.Marshal(raw)
 			require.NoError(t, err)
-			err = validate(bytes)
+			err = validate(bytes, raw.schemaLoader())
 			require.Error(t, err)
 			require.Contains(t, err.Error(), "serviceEndpoint is required")
 		}
@@ -638,7 +645,7 @@ func TestValidateDidDocCreated(t *testing.T) {
 			raw.Created = nil
 			bytes, err := json.Marshal(raw)
 			require.NoError(t, err)
-			err = validate(bytes)
+			err = validate(bytes, raw.schemaLoader())
 			require.NoError(t, err)
 		}
 	})
@@ -662,7 +669,7 @@ func TestValidateDidDocUpdated(t *testing.T) {
 			raw.Updated = nil
 			bytes, err := json.Marshal(raw)
 			require.NoError(t, err)
-			err = validate(bytes)
+			err = validate(bytes, raw.schemaLoader())
 			require.NoError(t, err)
 		}
 	})
@@ -686,7 +693,7 @@ func TestValidateDidDocProof(t *testing.T) {
 			raw.Proof = nil
 			bytes, err := json.Marshal(raw)
 			require.NoError(t, err)
-			err = validate(bytes)
+			err = validate(bytes, raw.schemaLoader())
 			require.NoError(t, err)
 		}
 	})
@@ -701,7 +708,7 @@ func TestValidateDidDocProof(t *testing.T) {
 			delete(proof, jsonldType)
 			bytes, err := json.Marshal(raw)
 			require.NoError(t, err)
-			err = validate(bytes)
+			err = validate(bytes, raw.schemaLoader())
 			require.Error(t, err)
 			require.Contains(t, err.Error(), "type is required")
 		}
@@ -717,7 +724,7 @@ func TestValidateDidDocProof(t *testing.T) {
 			delete(proof, jsonldCreated)
 			bytes, err := json.Marshal(raw)
 			require.NoError(t, err)
-			err = validate(bytes)
+			err = validate(bytes, raw.schemaLoader())
 			require.Error(t, err)
 			require.Contains(t, err.Error(), "created is required")
 		}
@@ -733,7 +740,7 @@ func TestValidateDidDocProof(t *testing.T) {
 			delete(proof, jsonldCreator)
 			bytes, err := json.Marshal(raw)
 			require.NoError(t, err)
-			err = validate(bytes)
+			err = validate(bytes, raw.schemaLoader())
 			require.Error(t, err)
 			require.Contains(t, err.Error(), "creator is required")
 		}
@@ -747,7 +754,7 @@ func TestValidateDidDocProof(t *testing.T) {
 		delete(proof, jsonldProofValue)
 		bytes, err := json.Marshal(raw)
 		require.NoError(t, err)
-		err = validate(bytes)
+		err = validate(bytes, raw.schemaLoader())
 		require.Error(t, err)
 		require.Contains(t, err.Error(), "proofValue is required")
 	})
@@ -760,7 +767,7 @@ func TestValidateDidDocProof(t *testing.T) {
 		delete(proof, jsonldSignatureValue)
 		bytes, err := json.Marshal(raw)
 		require.NoError(t, err)
-		err = validate(bytes)
+		err = validate(bytes, raw.schemaLoader())
 		require.Error(t, err)
 		require.Contains(t, err.Error(), "signatureValue is required")
 	})
@@ -775,7 +782,7 @@ func TestValidateDidDocProof(t *testing.T) {
 			delete(proof, jsonldDomain)
 			bytes, err := json.Marshal(raw)
 			require.NoError(t, err)
-			err = validate(bytes)
+			err = validate(bytes, raw.schemaLoader())
 			require.NoError(t, err)
 		}
 	})
@@ -790,7 +797,7 @@ func TestValidateDidDocProof(t *testing.T) {
 			delete(proof, jsonldNonce)
 			bytes, err := json.Marshal(raw)
 			require.NoError(t, err)
-			err = validate(bytes)
+			err = validate(bytes, raw.schemaLoader())
 			require.NoError(t, err)
 		}
 	})

